### PR TITLE
Multiple outputs with deps

### PR DESCRIPTION
**Pull request by moroten**
This PR is a rebase https://github.com/ninja-build/ninja/pull/1236 with more tests added. When adding the tests, I realised that deps for the implicit outputs also have to be recorded.

### CODE REVIEW

File: src/build_test.cc [hunk 1]
```diff
@@ -1882,6 +1882,308 @@ TEST_F(BuildWithDepsLogTest, Straightforward) {
...
+
+    // We should have rebuilt the output due to in2 being
+    // out of date.
+    EXPECT_EQ(1u, command_runner_.commands_ran_.size());
+
+    builder.command_runner_.release();
+  }
+}
+
+/// Verify that the deps are recorded for each output output.
```
Comment (nico) [inline src/build_test.cc]: nit: output output

File: src/manifest_parser.cc [hunk 2]
```diff
@@ -404,7 +404,7 @@ bool ManifestParser::ParseEdge(string* err) {
 
   // Multiple outputs aren't (yet?) supported with depslog.
```
Comment (nico) [inline src/manifest_parser.cc]: Update comment.

File: src/manifest_parser.cc [hunk 3]
```diff
@@ -404,7 +404,7 @@ bool ManifestParser::ParseEdge(string* err) {
...
   string deps_type = edge->GetBinding("deps");
-  if (!deps_type.empty() && edge->outputs_.size() > 1) {
+  if (!deps_type.empty() && edge->outputs_.size() - edge->implicit_outs_ != 1) {
     return lexer_.Error("multiple outputs aren't (yet?) supported by depslog; "
```
Comment (nico) [inline src/manifest_parser.cc]: Maybe say "non-implicit outputs" here too.

**Comment by nico** [comment]
Thanks much for working on this!

The approach here is to write a full deps entry with all inputs for each implicit output (meaning for an edge with N outputs and M inputs, you store N*M ints). This is simple, which is good.

Can you update the description to say "Multiple implicit outputs"? Multiple real outputs still aren't supported, right? Could this approach work for multiple non-implicit outputs as well?

What will happen if ninja is killed when it has written deps for some of the outputs but not all of them (both if old deps are in the log and if not)? I think this will effectively take the union of old and new deps for the edge, which should do the right thing (?).

It looks @danw didn't think of adding implicit outputs as necessary – maybe he has thoughts on the addition?

And @evmar has designed the deps log and always comes up with corner cases I can't think of.

**Comment by moroten** [comment]
Regarding multiple real/non-implicit outputs: What is the logic of the .d file? My assumption is that the .d file should have one entry, which points to the single real output, and that all the implicit outputs just inherits that.

Supporting multiple entries in the .d file would require a lot more work and probably break a lot of setups as there is currently no check what the entry in the .d file is called.

Regarting writing partial deps log, I think, but am not sure, that `LoadDepsFromLog()` will check the mtime recorded in the deps log and conclude that the deps information is invalid and therefore rebuild the target.

I will update the pull request later when all your questions has got a response.

**Comment by nico** [comment]
To be clear, I'm not asking you to add support for multiple outputs. I agree it'd require depfile parser changes and probably changes all over the place; I'm just curious if you think that the deps log part of such a change could work with this approach here.

I'll wait a bit if danw or evmar want to comment, but if they don't I'll take this patch (with the minor comments above addressed). Thanks again!

**Comment by danw** [comment]
I don't think it's harmful, my approach would have just preferred to rebuild if the the structure of the outputs of the rule changed that much (we've run into occasional incremental build bugs due to inputs/outputs not being listed directly in the command line, so not being hashed)

Related to this discussion, I've discovered that the protobuf compiler (protoc) writes out a dependency file with multiple outputs, at least when generating C bindings. My current plan is just to rewrite the dep file using sed before giving it back to ninja.

**Comment by mathstuf** [comment]
How does this interact with #1364 (sorry, I missed this when I implemented it)? It certainly conflicts, but the implicit output side might not be taken into account in that PR.

**Comment by moroten** [comment]
> How does this interact with #1364 (sorry, I missed this when I implemented it)? It certainly conflicts, but the implicit output side might not be taken into account in that PR.

This is #1364, I don't understand your concerns.

**Comment by mathstuf** [comment]
Sorry, I meant #1534.

**Comment by moroten** [comment]
I am not very active any more in the Ninja development as I am moving to Bazel. Please feel free to pick the best from both pull requests (#1364 and #1534).

**Comment by jhasse** [comment]
Closing for now until someone wants to pick it up.

**Pull request closed without merging**